File: auto-translate.js
#!/usr/bin/env node
/**
 * 自动翻译 Markdown 文档脚本
 * 使用简单的规则和映射来翻译文档
 */

const fs = require('fs');
const path = require('path');

// 需要翻译的文件列表
const filesToTranslate = [
  'extension-guides/ai/chat.md',
  'extension-guides/ai/language-model-chat-provider.md',
  'extension-guides/ai/language-model-tutorial.md',
  'extension-guides/ai/language-model.md',
  'extension-guides/ai/mcp.md',
  'extension-guides/ai/tools.md',
  'extension-guides/color-theme.md',
  'extension-guides/custom-editors.md',
  'extension-guides/markdown-extension.md',
  'extension-guides/notebook.md',
  'extension-guides/overview.md',
  'extension-guides/product-icon-theme.md',
  'extension-guides/scm-provider.md',
  'extension-guides/task-provider.md',
  'extension-guides/telemetry.md',
  'extension-guides/testing.md',
  'extension-guides/tree-view.md',
  'extension-guides/virtual-documents.md',
  'extension-guides/virtual-workspaces.md',
  'extension-guides/web-extensions.md',
  'extension-guides/webview.md',
  'extension-guides/workspace-trust.md',
  'references/activation-events.md',
  'references/commands.md',
  'references/contribution-points.md',
  'references/document-selector.md',
  'references/extension-manifest.md',
  'references/icons-in-labels.md',
  'references/theme-color.md',
  'references/when-clause-contexts.md',
  'ux-guidelines/notifications.md',
  'ux-guidelines/panel.md',
  'ux-guidelines/quick-picks.md',
  'ux-guidelines/settings.md',
  'ux-guidelines/sidebars.md',
  'ux-guidelines/status-bar.md',
  'ux-guidelines/views.md',
  'ux-guidelines/walkthroughs.md',
  'ux-guidelines/webviews.md',
  'working-with-extensions/continuous-integration.md',
  'working-with-extensions/publishing-extension.md',
  'working-with-extensions/testing-extension.md',
];

// 常见英文短语到中文的映射
const translations = {
  // 文档结构相关
  'This guide': '本指南',
  'This topic': '本主题',
  'This section': '本节',
  'This article': '本文',
  'This page': '本页',
  'This document': '本文档',
  
  // 指示词
  'For example': '例如',
  'For instance': '例如',
  'For more information': '有关更多信息',
  'For details': '详情请见',
  'To learn more': '要了解更多',
  'Learn more': '了解更多',
  'See also': '另请参阅',
  'Refer to': '请参阅',
  'Check out': '查看',
  
  // 操作相关
  'You can': '你可以',
  'You should': '你应该',
  'You need to': '你需要',
  'You must': '你必须',
  'To do this': '要执行此操作',
  'In order to': '为了',
  
  // 提示词
  'Note:': '注意：',
  'Tip:': '提示：',
  'Warning:': '警告：',
  'Important:': '重要：',
  'Example:': '示例：',
  
  // 技术术语 (保留一些，翻译一些)
  'extension': '插件',
  'Extension': '插件',
  'extensions': '插件',
  'Extensions': '插件',
  'plugin': '插件',
  'Plugin': '插件',
  'add-on': '插件',
  'Add-on': '插件',
  
  'workspace': '工作区',
  'Workspace': '工作区',
  'workspaces': '工作区',
  'Workspaces': '工作区',
  
  'command': '命令',
  'Command': '命令',
  'commands': '命令',
  'Commands': '命令',
  
  'configuration': '配置',
  'Configuration': '配置',
  'config': '配置',
  'Config': '配置',
  'settings': '设置',
  'Settings': '设置',
  
  'debug': '调试',
  'Debug': '调试',
  'debugger': '调试器',
  'Debugger': '调试器',
  'debugging': '调试',
  'Debugging': '调试',
  
  'workbench': '工作台',
  'Workbench': '工作台',
  
  'theme': '主题',
  'Theme': '主题',
  'themes': '主题',
  'Themes': '主题',
  
  'marketplace': '市场',
  'Marketplace': '市场',
  
  'activation event': '激活事件',
  'Activation Event': '激活事件',
  'activation events': '激活事件',
  'Activation Events': '激活事件',
  
  'contribution point': '贡献点',
  'Contribution Point': '贡献点',
  'contribution points': '贡献点',
  'Contribution Points': '贡献点',
  
  'manifest': '清单',
  'Manifest': '清单',
  
  'provider': '提供器',
  'Provider': '提供器',
  'providers': '提供器',
  'Providers': '提供器',
  
  'editor': '编辑器',
  'Editor': '编辑器',
  'editors': '编辑器',
  'Editors': '编辑器',
  
  'status bar': '状态栏',
  'Status Bar': '状态栏',
  'statusbar': '状态栏',
  'StatusBar': '状态栏',
  
  'sidebar': '侧边栏',
  'Sidebar': '侧边栏',
  'side bar': '侧边栏',
  'Side Bar': '侧边栏',
  
  'panel': '面板',
  'Panel': '面板',
  'panels': '面板',
  'Panels': '面板',
  
  'terminal': '终端',
  'Terminal': '终端',
  
  'notification': '通知',
  'Notification': '通知',
  'notifications': '通知',
  'Notifications': '通知',
  
  'hover': '悬停',
  'Hover': '悬停',
  
  'completion': '补全',
  'Completion': '补全',
  'completions': '补全',
  'Completions': '补全',
  'autocomplete': '自动补全',
  'AutoComplete': '自动补全',
  
  'diagnostic': '诊断',
  'Diagnostic': '诊断',
  'diagnostics': '诊断',
  'Diagnostics': '诊断',
};

function main() {
  const baseDir = path.join(__dirname, 'docs', 'plugin-docs');
  
  console.log('自动翻译 Markdown 文档');
  console.log('========================\n');
  console.log(`基础目录: ${baseDir}`);
  console.log(`需要翻译的文件数: ${filesToTranslate.length}\n`);
  
  let successCount = 0;
  let failCount = 0;
  
  for (const relPath of filesToTranslate) {
    const filePath = path.join(baseDir, relPath);
    
    try {
      if (!fs.existsSync(filePath)) {
        console.log(`⚠️  跳过: ${relPath} (文件不存在)`);
        failCount++;
        continue;
      }
      
      // 读取文件
      const content = fs.readFileSync(filePath, 'utf-8');
      
      // 简单替换（注意：这是一个非常简化的翻译，实际需要更复杂的处理）
      let translated = content;
      
      // 在代码块外进行替换
      const lines = content.split('\n');
      let inCodeBlock = false;
      let inFrontmatter = false;
      let result = [];
      
      for (let i = 0; i < lines.length; i++) {
        let line = lines[i];
        
        // 检测 frontmatter
        if (i === 0 && line.trim() === '---') {
          inFrontmatter = true;
          result.push(line);
          continue;
        }
        if (inFrontmatter && line.trim() === '---') {
          inFrontmatter = false;
          result.push(line);
          continue;
        }
        
        // 在 frontmatter 中，跳过翻译
        if (inFrontmatter) {
          result.push(line);
          continue;
        }
        
        // 检测代码块
        if (line.trim().startsWith('```')) {
          inCodeBlock = !inCodeBlock;
          result.push(line);
          continue;
        }
        
        // 在代码块中，跳过翻译
        if (inCodeBlock) {
          result.push(line);
          continue;
        }
        
        // 执行简单的替换翻译
        for (const [eng, chn] of Object.entries(translations)) {
          // 使用单词边界来避免部分匹配
          const regex = new RegExp(`\\b${eng.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\b`, 'g');
          line = line.replace(regex, chn);
        }
        
        result.push(line);
      }
      
      translated = result.join('\n');
      
      // 如果内容有变化，写回文件
      if (translated !== content) {
        fs.writeFileSync(filePath, translated, 'utf-8');
        console.log(`✅ 已翻译: ${relPath}`);
        successCount++;
      } else {
        console.log(`⏭️  跳过: ${relPath} (无需翻译)`);
      }
      
    } catch (error) {
      console.log(`❌ 失败: ${relPath}`);
      console.log(`   错误: ${error.message}`);
      failCount++;
    }
  }
  
  console.log(`\n翻译完成!`);
  console.log(`✅ 成功: ${successCount}`);
  console.log(`❌ 失败: ${failCount}`);
  console.log(`⏭️  跳过: ${filesToTranslate.length - successCount - failCount}`);
}

if (require.main === module) {
  main();
}

File: auto_translate.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
自动翻译 Markdown 文档
使用简单的规则和替换来翻译文档
"""

import os
import re
from pathlib import Path
import json

# VSCode -> Baosky 替换
BRAND_REPLACEMENTS = {
    'VSCode': 'Baosky',
    'VS Code': 'Baosky',
    'vscode': 'baosky',
    'Visual Studio Code': 'Baosky'
}

# Extension -> 插件 替换（需要智能处理）
TERM_REPLACEMENTS = {
    'extension': '插件',
    'Extension': '插件',
    'EXTENSION': '插件',
    'extensions': '插件',
    'Extensions': '插件',
}

# 常见词汇翻译
COMMON_TRANSLATIONS = {
    # 基础术语
    'Overview': '概述',
    'Getting Started': '快速入门',
    'Get Started': '快速入门',
    'Introduction': '介绍',
    'Guide': '指南',
    'Tutorial': '教程',
    'Reference': '参考',
    'API': 'API',
    'Example': '示例',
    'Examples': '示例',

    # VS Code 特定术语
    'workspace': '工作区',
    'Workspace': '工作区',
    'workbench': '工作台',
    'Workbench': '工作台',
    'editor': '编辑器',
    'Editor': '编辑器',
    'command': '命令',
    'Command': '命令',
    'commands': '命令',
    'Commands': '命令',
    'theme': '主题',
    'Theme': '主题',
    'themes': '主题',
    'Themes': '主题',
    'debug': '调试',
    'Debug': '调试',
    'debugging': '调试',
    'Debugging': '调试',
    'debugger': '调试器',
    'Debugger': '调试器',
    'configuration': '配置',
    'Configuration': '配置',
    'settings': '设置',
    'Settings': '设置',
    'marketplace': '市场',
    'Marketplace': '市场',
    'package': '包',
    'Package': '包',
    'manifest': '清单',
    'Manifest': '清单',
    'activation': '激活',
    'Activation': '激活',
    'contribution': '贡献',
    'Contribution': '贡献',
    'provider': '提供器',
    'Provider': '提供器',
    'language': '语言',
    'Language': '语言',
    'syntax': '语法',
    'Syntax': '语法',
    'semantic': '语义',
    'Semantic': '语义',
    'snippet': '代码片段',
    'Snippet': '代码片段',
    'snippets': '代码片段',
    'Snippets': '代码片段',
    'completion': '补全',
    'Completion': '补全',
    'IntelliSense': 'IntelliSense',
    'hover': '悬停',
    'Hover': '悬停',
    'definition': '定义',
    'Definition': '定义',
    'reference': '引用',
    'Reference': '引用',
    'references': '引用',
    'References': '引用',
    'symbol': '符号',
    'Symbol': '符号',
    'symbols': '符号',
    'Symbols': '符号',
    'diagnostic': '诊断',
    'Diagnostic': '诊断',
    'diagnostics': '诊断',
    'Diagnostics': '诊断',
    'notification': '通知',
    'Notification': '通知',
    'notifications': '通知',
    'Notifications': '通知',
    'status bar': '状态栏',
    'Status Bar': '状态栏',
    'activity bar': '活动栏',
    'Activity Bar': '活动栏',
    'sidebar': '侧边栏',
    'Sidebar': '侧边栏',
    'panel': '面板',
    'Panel': '面板',
    'terminal': '终端',
    'Terminal': '终端',
    'output': '输出',
    'Output': '输出',
    'task': '任务',
    'Task': '任务',
    'tasks': '任务',
    'Tasks': '任务',
    'test': '测试',
    'Test': '测试',
    'testing': '测试',
    'Testing': '测试',
    'publish': '发布',
    'Publish': '发布',
    'publishing': '发布',
    'Publishing': '发布',
    'bundling': '打包',
    'Bundling': '打包',
    'migration': '迁移',
    'Migration': '迁移',
    'template': '模板',
    'Template': '模板',
    'webview': 'Webview',
    'Webview': 'Webview',
    'tree view': '树视图',
    'Tree View': '树视图',
    'custom editor': '自定义编辑器',
    'Custom Editor': '自定义编辑器',
    'icon': '图标',
    'Icon': '图标',
    'icons': '图标',
    'Icons': '图标',
    'color': '颜色',
    'Color': '颜色',
    'colors': '颜色',
    'Colors': '颜色',
    'file': '文件',
    'File': '文件',
    'files': '文件',
    'Files': '文件',
    'folder': '文件夹',
    'Folder': '文件夹',
    'folders': '文件夹',
    'Folders': '文件夹',
    'document': '文档',
    'Document': '文档',
    'documents': '文档',
    'Documents': '文档',
    'text': '文本',
    'Text': '文本',
    'content': '内容',
    'Content': '内容',
    'user': '用户',
    'User': '用户',
    'input': '输入',
    'Input': '输入',
    'quick pick': '快速选择',
    'Quick Pick': '快速选择',
    'context menu': '上下文菜单',
    'Context Menu': '上下文菜单',
    'menu': '菜单',
    'Menu': '菜单',
    'keybinding': '快捷键',
    'Keybinding': '快捷键',
    'keybindings': '快捷键',
    'Keybindings': '快捷键',
    'keyboard shortcut': '键盘快捷键',
    'Keyboard Shortcut': '键盘快捷键',
}

def is_in_code_block(lines, line_index):
    """检查当前行是否在代码块中"""
    code_block_count = 0
    for i in range(line_index):
        if lines[i].strip().startswith('```'):
            code_block_count += 1
    return code_block_count % 2 == 1

def is_in_frontmatter(lines, line_index):
    """检查当前行是否在 frontmatter 中"""
    dash_count = 0
    for i in range(line_index + 1):
        if lines[i].strip() == '---':
            dash_count += 1
            if dash_count == 2:
                return i >= line_index
    return dash_count == 1 and line_index >= 0

def preserve_markdown_elements(text):
    """保护 markdown 元素不被翻译"""
    # 保护图片（先处理图片，因为图片也包含 []() 模式）
    images = re.findall(r'!\[([^\]]*)\]\(([^)]+)\)', text)
    image_placeholders = []
    for i, (alt_text, img_url) in enumerate(images):
        original = f'![{alt_text}]({img_url})'
        placeholder = f'__IMAGE_{i}__'
        text = text.replace(original, placeholder, 1)
        image_placeholders.append((placeholder, original))

    # 保护链接
    links = re.findall(r'\[([^\]]+)\]\(([^)]+)\)', text)
    link_placeholders = []
    for i, (link_text, link_url) in enumerate(links):
        original = f'[{link_text}]({link_url})'
        placeholder = f'__LINK_{i}__'
        text = text.replace(original, placeholder, 1)
        link_placeholders.append((placeholder, original))

    # 保护内联代码
    inline_codes = re.findall(r'`[^`]+`', text)
    code_placeholders = []
    for i, code in enumerate(inline_codes):
        placeholder = f'__INLINE_CODE_{i}__'
        text = text.replace(code, placeholder, 1)
        code_placeholders.append((placeholder, code))

    return text, code_placeholders, link_placeholders, image_placeholders

def restore_markdown_elements(text, code_placeholders, link_placeholders, image_placeholders):
    """恢复 markdown 元素"""
    # 恢复内联代码
    for placeholder, code in code_placeholders:
        text = text.replace(placeholder, code)

    # 恢复链接
    for placeholder, original in link_placeholders:
        text = text.replace(placeholder, original)

    # 恢复图片
    for placeholder, original in image_placeholders:
        text = text.replace(placeholder, original)

    return text

def translate_line(line):
    """翻译单行文本"""
    if not line.strip():
        return line

    # 保护 markdown 元素
    original_line = line
    protected_line, code_placeholders, link_placeholders, image_placeholders = preserve_markdown_elements(line)

    # 替换品牌名称
    for en, zh in BRAND_REPLACEMENTS.items():
        protected_line = protected_line.replace(en, zh)

    # 替换术语（需要智能处理，避免替换 URL 中的词）
    # 这里简单处理，只替换单词边界的术语
    for en, zh in TERM_REPLACEMENTS.items():
        # 使用单词边界匹配
        protected_line = re.sub(r'\b' + en + r'\b', zh, protected_line)

    # 恢复 markdown 元素
    translated_line = restore_markdown_elements(protected_line, code_placeholders, link_placeholders, image_placeholders)

    return translated_line

def process_file(file_path):
    """处理单个文件"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        lines = content.split('\n')
        new_lines = []

        in_code_block = False
        in_frontmatter = False
        frontmatter_count = 0

        for i, line in enumerate(lines):
            # 检查 frontmatter
            if line.strip() == '---':
                frontmatter_count += 1
                if frontmatter_count <= 2:
                    in_frontmatter = frontmatter_count == 1
                new_lines.append(line)
                continue

            # 在 frontmatter 中，只处理 MetaDescription
            if in_frontmatter:
                if 'MetaDescription:' in line:
                    # 翻译 MetaDescription
                    new_lines.append(translate_line(line))
                else:
                    new_lines.append(line)
                continue

            # 检查代码块
            if line.strip().startswith('```'):
                in_code_block = not in_code_block
                new_lines.append(line)
                continue

            # 在代码块中，不翻译
            if in_code_block:
                new_lines.append(line)
                continue

            # 翻译普通文本行
            translated = translate_line(line)
            new_lines.append(translated)

        # 写回文件
        new_content = '\n'.join(new_lines)
        if new_content != content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(new_content)
            return True
        return False

    except Exception as e:
        print(f"错误处理文件 {file_path}: {e}")
        return False

def process_directory(directory):
    """处理目录中的所有 md 文件"""
    processed_count = 0
    total_count = 0

    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith('.md'):
                total_count += 1
                file_path = os.path.join(root, file)
                rel_path = os.path.relpath(file_path, directory)
                print(f"处理: {rel_path}")

                if process_file(file_path):
                    processed_count += 1
                    print(f"  ✓ 已更新")
                else:
                    print(f"  - 无更改")

    return processed_count, total_count

def main():
    """主函数"""
    base_dir = Path("/root/my_website/my-website/docs/plugin-docs")

    if not base_dir.exists():
        print(f"错误：目录不存在: {base_dir}")
        return

    print("开始自动翻译文档...\n")
    print(f"目标目录: {base_dir}\n")

    processed, total = process_directory(base_dir)

    print(f"\n{'='*50}")
    print(f"翻译完成!")
    print(f"总文件数: {total}")
    print(f"已处理: {processed}")
    print(f"{'='*50}")

if __name__ == "__main__":
    main()
